File: src/sources/local.rs
use super::{ProgressReporter, Source};
use std::path::{Path, PathBuf};

/// Passthrough source for local filesystem paths.
pub struct LocalSource {
    pub path: String,
}

#[async_trait::async_trait]
impl Source for LocalSource {
    async fn sync_to_local(
        &self,
        _staging_root: &Path,
        progress: &mut dyn ProgressReporter,
    ) -> Result<PathBuf, Box<dyn std::error::Error + Send + Sync>> {
        let p = Path::new(&self.path);
        let resolved = if p.is_absolute() {
            p.to_path_buf()
        } else {
            std::env::current_dir()?.join(p)
        };
        let canonical = std::fs::canonicalize(&resolved).unwrap_or(resolved);
        progress.report(&format!("local source at {}", canonical.display()));
        Ok(canonical)
    }

    fn name(&self) -> &str {
        "local"
    }
}

File: go/internal/sources/local.go
// Local filesystem source (Rust src/sources/local.rs): a passthrough that
// resolves the path and hands it to the indexer.
package sources

import (
	"context"
	"os"
	"path/filepath"
)

// LocalSource resolves a local filesystem path. It never copies anything: the
// staging root is ignored, exactly like the Rust implementation.
type LocalSource struct {
	Path string
}

// Name implements Source.
func (l *LocalSource) Name() string { return "local" }

// SyncToLocal resolves Path against the working directory and canonicalizes it
// (Rust canonicalize with a fallback to the joined path when the path does not
// fully resolve, e.g. a dangling symlink target).
func (l *LocalSource) SyncToLocal(_ context.Context, _ string, progress ProgressReporter) (string, error) {
	resolved := l.Path
	if !filepath.IsAbs(resolved) {
		cwd, err := os.Getwd()
		if err != nil {
			return "", err
		}
		resolved = filepath.Join(cwd, resolved)
	}
	canonical := resolved
	if real, err := filepath.EvalSymlinks(resolved); err == nil {
		canonical = real
	}
	report(progress, "local source at "+canonical)
	return canonical, nil
}

// RemoteFingerprint returns "" : a local path cannot be polled (Rust's default
// trait implementation returns Ok(None)).
func (l *LocalSource) RemoteFingerprint(context.Context) (string, error) { return "", nil }

// MaterializeEphemeral delegates to SyncToLocal (Rust's trait default).
func (l *LocalSource) MaterializeEphemeral(ctx context.Context, stagingRoot string, progress ProgressReporter) (string, error) {
	return l.SyncToLocal(ctx, stagingRoot, progress)
}
